util/htpasswd/htpasswd_test.go

```diff
@@ -0,0 +1,224 @@
+package htpasswd
+
+import (
+	"context"
+	"encoding/base64"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc/metadata"
+)
+
+func TestParseHtpasswdAuth(t *testing.T) {
+	tests := []struct {
+		input           string
+		user            string
+		encodedPassword string
+	}{
+		{
+			input:           "user:$2y$10$q6I6fxG2c79jBSXJ8L2jde15czipSRpu/uhW5Le.ooJLyfXiaPDZG",
+			user:            "user",
+			encodedPassword: "$2y$10$q6I6fxG2c79jBSXJ8L2jde15czipSRpu/uhW5Le.ooJLyfXiaPDZG",
+		},
+		{
+			input:           "user1:$2y$10$/4EcZtrJUgivhcTJPGOz/uhQEUAQP.zvThFwIHwdjQT97iL4gWMri",
+			user:            "user1",
+			encodedPassword: "$2y$10$/4EcZtrJUgivhcTJPGOz/uhQEUAQP.zvThFwIHwdjQT97iL4gWMri",
+		},
+		{
+			input:           "user2:$2y$10$xXmx6BQv6re3P2sOAoPGNu/MJOwWxDtxtNzlEJ2qkUVRK6SqAXD9m",
+			user:            "user2",
+			encodedPassword: "$2y$10$xXmx6BQv6re3P2sOAoPGNu/MJOwWxDtxtNzlEJ2qkUVRK6SqAXD9m",
+		},
+		{
+			input:           "user3:$2y$10$eKLWzld7iMPrcyDqam8.Y.R1deeSUBWFD3P6eQHJ0Iqa1qR4yBxaq",
+			user:            "user3",
+			encodedPassword: "$2y$10$eKLWzld7iMPrcyDqam8.Y.R1deeSUBWFD3P6eQHJ0Iqa1qR4yBxaq",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.user, func(t *testing.T) {
+			user, encodedPass, err := ParseHtpasswdAuth(tt.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			assert.Equal(t, user+":"+encodedPass, tt.input)
+		})
+	}
+}
+
+func TestCompareBasicAuth(t *testing.T) {
+	tests := []struct {
+		name        string
+		input       string
+		user        string
+		password    string
+		expectError bool
+	}{
+		{
+			name:        "SuccessfulAuthentication",
+			input:       "user:$2y$10$q6I6fxG2c79jBSXJ8L2jde15czipSRpu/uhW5Le.ooJLyfXiaPDZG", // hashed 'foobar'
+			user:        "user",
+			password:    "foobar",
+			expectError: false,
+		},
+		{
+			name:        "UserMismatch",
+			input:       "user:$2y$10$q6I6fxG2c79jBSXJ8L2jde15czipSRpu/uhW5Le.ooJLyfXiaPDZG",
+			user:        "wronguser",
+			password:    "foobar",
+			expectError: true,
+		},
+		{
+			name:        "PasswordMismatch",
+			input:       "user:$2y$10$q6I6fxG2c79jBSXJ8L2jde15czipSRpu/uhW5Le.ooJLyfXiaPDZG",
+			user:        "user",
+			password:    "wrongpassword",
+			expectError: true,
+		},
+		{
+			name:        "MalformedCredential",
+			input:       "malformed",
+			user:        "user",
+			password:    "foobar",
+			expectError: true,
```

Comment:
Specify which error is expected. If it is `nil`, then it should pass without error. Specifying the error type helps ensure that the proper error is captured.